### PR TITLE
libbeat/kibana: add ClientConfig.Headers

### PR DESCRIPTION
**Pull request by axw**
## What does this PR do?

Add configurable headers to include in every request sent to Kibana.

## Why is it important?

This is needed in some security applications, for authenticating the Beat. Specifically, this is needed to fix https://github.com/elastic/apm-server/issues/4065

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~

## How to test this PR locally

1. Run `nc -l 1234`
2. Configure filebeat with `setup.kibana.host: localhost:1234`, and a custom header in `setup.kibana.headers`
3. Run `filebeat setup`

Observe the custom header is logged by netcat:

<pre>
GET /api/status HTTP/1.1
Host: localhost:1234
User-Agent: Go-http-client/1.1
Accept: application/json
Content-Type: application/json
<b>Foo: bar</b>
Kbn-Xsrf: 1
Accept-Encoding: gzip
</pre>

## Related issues

https://github.com/elastic/apm-server/issues/4065

## Use cases

Introduces a means of specifying custom HTTP headers to add to all requests sent to Kibana. As above, needed in some security use cases.

### CODE REVIEW

File: libbeat/kibana/client.go [hunk 1]
```diff
@@ -206,14 +213,18 @@ func (conn *Connection) SendWithContext(ctx context.Context, method, extraPath s
 	req.Header.Set("Content-Type", "application/json")
 	req.Header.Add("Accept", "application/json")
 	req.Header.Set("kbn-xsrf", "1")
+	addHeaders(req.Header, conn.Headers)
+	addHeaders(req.Header, headers)
```
Comment (simitt) [inline libbeat/kibana/client.go:217]: nit: Should the `conn.Headers`  and `headers` be added before the `Content-Type` and `kbn-xsrf` are `set`?

Comment (axw) [reply libbeat/kibana/client.go:217]: Good question. I'm not sure - I will defer to the Beats team. `headers` was already set after, so I left it there.

File: x-pack/heartbeat/heartbeat.reference.yml [hunk 2]
```diff
@@ -758,11 +758,11 @@ output.elasticsearch:
   # Resolve names locally when using a proxy server. Defaults to false.
   #proxy_use_local_resolver: false
 
-  # Enable SSL support. SSL is automatically enabled if any SSL setting is set.
+  # Use SSL settings for HTTPS.
```
Comment (axw) [inline x-pack/heartbeat/heartbeat.reference.yml:761]: It appears that this file is out of date on master. These changes show up after `make update` on master too.

File: libbeat/kibana/client.go (hunk 1 above)
Comment (urso) [reply libbeat/kibana/client.go:217]: I think I'd prefer to call addHeaders before setting the contents type. These are hard coded and should not be modifiable. well, in the end it doesn't really matter for most users.

Comment (axw) [reply libbeat/kibana/client.go:217]: That sounds fine to me, thanks for weighing in. I'll update and add a test that checks that.

File: x-pack/heartbeat/heartbeat.reference.yml (hunk 2 above)
Comment (urso) [reply x-pack/heartbeat/heartbeat.reference.yml:761]: Hm, let me check that and open another PR if we have some problems. This file should be fixed ASAP.

Comment (jsoriano) [reply x-pack/heartbeat/heartbeat.reference.yml:761]: This should be fixed by https://github.com/elastic/beats/pull/20681

Comment (axw) [reply x-pack/heartbeat/heartbeat.reference.yml:761]: I have also updated this in https://github.com/elastic/beats/pull/20683, which I plan to merge as soon as the tests pass

File: libbeat/kibana/client.go (hunk 1 above)
Comment (axw) [reply libbeat/kibana/client.go:214]: Updated to set in order: headers configured in config, then in the function call, then statically defined ones (Content-Type, Accept, kbn-xsrf).

Note that I changed `req.Header.Add("Accept", ...` to `req.Header.Set("Accept", ...`, which means the header is always just one value.

**Review by urso — APPROVED** [review]

**Comment by axw** [comment]
Test failures are unrelated AFAICT.

**Pull request merged**